packages/playwright-core/src/utils/traceUtils.ts

```diff
@@ -92,3 +95,51 @@ export async function mergeTraceFiles(fileName: string, temporaryTraceFiles: str
   });
   await mergePromise;
 }
+
+export async function saveTraceFile(fileName: string, traceEvents: ActionTraceEvent[], saveSources: boolean) {
+  const lines: string[] = traceEvents.map(e => JSON.stringify(e));
+  const zipFile = new yazl.ZipFile();
+  zipFile.addBuffer(Buffer.from(lines.join('\n')), 'trace.trace');
+
+  if (saveSources) {
+    const sourceFiles = new Set<string>();
+    for (const event of traceEvents) {
+      for (const frame of event.stack || [])
+        sourceFiles.add(frame.file);
+    }
+    for (const sourceFile of sourceFiles) {
+      const source = await fs.promises.readFile(sourceFile, 'utf8');
+      zipFile.addBuffer(Buffer.from(source), 'resources/src@' + calculateSha1(sourceFile) + '.txt');
+    }
+  }
+  await new Promise(f => {
+    zipFile.end(undefined, () => {
+      zipFile.outputStream.pipe(fs.createWriteStream(fileName)).on('close', f);
+    });
+  });
+}
+
+export function createTraceEventForExpect(apiName: string, expected: any, stack: StackFrame[], wallTime: number): ActionTraceEvent {
+  return {
+    type: 'action',
+    callId: 'expect@' + wallTime,
+    wallTime,
+    startTime: monotonicTime(),
+    endTime: 0,
+    class: 'Test',
+    method: 'step',
+    apiName,
+    params: { 'expected': bestEffortJsonStringify(expected) },
+    snapshots: [],
+    log: [],
+    stack,
+  };
+}
+
+function bestEffortJsonStringify(object: any): string {
+  try {
+    return JSON.stringify(object);
```

Comment:
`Buffer`s are stringifiable, so any `toMatchSnapshot` call will be very large. Let's skip them?